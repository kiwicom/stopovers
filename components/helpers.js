// @flow

import { scroller } from "react-scroll";

export function scrollToElement(elementName: string, callback?: () => void) {
  scroller.scrollTo(elementName, {
    smooth: true,
  });
  if (callback) {
    callback();
  }
}

export default { scrollToElement };
