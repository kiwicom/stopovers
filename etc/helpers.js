// @flow

import "isomorphic-unfetch";
import pick from "lodash.pick";
import { type LangInfos, type LangInfo } from "@kiwicom/nitro/lib/records/LangInfo";
import { type Language } from "@kiwicom/nitro/lib/records/Languages";
import { type BrandLanguages } from "@kiwicom/nitro/lib/records/BrandLanguage";
import cookies from "js-cookie";

export const usedLangIds = ["en", "cz", "ro", "hu", "es", "fr", "de", "ru", "it"];

export const UTM_PARAMS = process.env.UTM_PARAMS ? `?${process.env.UTM_PARAMS}` : "";

export function filterLanguages(langsData: LangInfos) {
  return pick(langsData, usedLangIds);
}

export function filterBrandLanguage(brandLangsData: BrandLanguages, langId: string) {
  const brandLanguage = brandLangsData.kiwicom[langId];
  const languages = pick(brandLanguage.languages, usedLangIds);
  return { ...brandLanguage, languages };
}

export function mapLanguage(lang: Language, langInfo: LangInfo) {
  return {
    ...langInfo,
    name: langInfo.displayName,
    flag: lang.flag || langInfo.id,
    defaultCountry: lang.defaultCountry,
  };
}

export function parseQuery(queryString: string): Object {
  const pairs = (queryString[0] === "?" ? queryString.substr(1) : queryString).split("&");
  return pairs.reduce((query, pair) => {
    const parts = pair.split("=");
    if (parts[0] !== "") {
      const key = decodeURIComponent(parts[0]);

      return {
        ...query,
        [key === "return" ? "returnDate" : key]: decodeURIComponent(parts[1] || ""),
      };
    }
    return query;
  }, {});
}

export function getCurrentUrlParams() {
  return parseQuery(window.location.search);
}

export function generateUserId() {
  function s4() {
    return Math.floor((1 + Math.random()) * 0x10000)
      .toString(16)
      .substring(1);
  }
  return `${s4() + s4()}-${s4()}-${s4()}-${s4()}-${s4()}${s4()}${s4()}`;
}

export const getUserId = () => {
  const userId = cookies.get("SKYPICKER_VISITOR_UNIQID");
  if (!userId) {
    const newId = generateUserId();
    cookies.set("SKYPICKER_VISITOR_UNIQID", newId, { expires: 365 * 4 });
    return newId;
  }
  return userId;
};
