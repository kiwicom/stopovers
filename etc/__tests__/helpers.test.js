// @flow

import Cookies from "js-cookie";

import * as Helpers from "../helpers";
import languages from "../../static/languages.json";

describe("helpers", () => {
  it("filters properties of an object properly", () => {
    const obj = {
      en: { iso: "en-GB" },
      ca: { iso: "en-CA" },
      us: { iso: "en-US" },
    };

    expect(Helpers.pick(obj, ["en"])).toEqual({ en: { iso: "en-GB" } });
    expect(Helpers.pick(obj, ["en-US", "en-CA"], "iso")).toEqual({
      ca: { iso: "en-CA" },
      us: { iso: "en-US" },
    });
  });

  it("filters languages properly", () => {
    const output = Helpers.filterLanguages(["en-GB", "en-CA", "en-US"]);
    expect(Object.keys(output)).toEqual(["en", "ca", "us"]);
    expect(Object.keys(output).every(key => output[key] === languages[key])).toBe(true);
  });

  it("gets correct language data", () => {
    expect(
      Helpers.getLanguage("non existing language", Helpers.filterLanguages(["en-GB"])),
    ).toEqual(languages.en);
    expect(Helpers.getLanguage("ca", Helpers.filterLanguages(["en-GB", "en-CA"]))).toEqual({
      ...languages.ca,
      name: languages.ca.displayName,
    });
    expect(Helpers.getLanguage("ca", Helpers.filterLanguages(["en-GB"]))).toEqual(languages.en);
    expect(Helpers.getLanguage("be", Helpers.filterLanguages(["en-GB", "fr-FR"]))).toEqual({
      ...languages.fr,
      name: languages.fr.displayName,
    });
  });

  it("gets correct brand language data", () => {
    const supportedLangs = Helpers.filterLanguages(["en-GB", "en-CA", "es-ES"]);
    expect(Helpers.getBrandLanguage("es", supportedLangs)).toMatchSnapshot();
  });

  it("gets correct args from current URL", () => {
    const paths = [
      "/some/path/?key=value",
      "/?a=an arg with spaces&b=followed by another arg",
      "/?symbols=an%20arg%20with%20encoded%20symbols%20%24%26%23%40",
    ];
    const outputs = paths.map((path, index) => {
      window.history.pushState({}, index, path);
      return Helpers.getCurrentUrlParams();
    });

    expect(outputs).toEqual([
      { key: "value" },
      { a: "an arg with spaces", b: "followed by another arg" },
      { symbols: "an arg with encoded symbols $&#@" },
    ]);
  });

  it("gets and sets user ID properly", () => {
    Cookies.set = jest.fn();
    Cookies.get = jest
      .fn()
      .mockImplementationOnce(() => "userID")
      .mockImplementationOnce(() => undefined);

    expect(Helpers.getUserId()).toEqual("userID");
    expect(Cookies.set).not.toBeCalled();

    const userID = Helpers.getUserId();
    expect(userID.length).toBe(36);
    expect(Cookies.set.mock.calls[0][0]).toEqual("SKYPICKER_VISITOR_UNIQID");
    expect(Cookies.set.mock.calls[0][1]).toEqual(userID);
  });
});
