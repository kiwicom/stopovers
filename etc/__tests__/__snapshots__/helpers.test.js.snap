// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`helpers gets correct brand language data 1`] = `
Object {
  "continents": Array [
    "ap",
    "eu",
    "aas",
    "mea",
  ],
  "defaultLocale": "en",
  "languages": Object {
    "ca": Object {
      "continent": "aas",
      "defaultCountry": "ca",
      "flag": "ca",
      "id": "ca",
      "name": "English (Canada)",
    },
    "en": Object {
      "continent": "eu",
      "defaultCountry": "gb",
      "flag": "gb",
      "id": "en",
      "name": "English",
    },
    "es": Object {
      "continent": "eu",
      "defaultCountry": "es",
      "flag": "es",
      "id": "es",
      "name": "Espa√±ol",
    },
  },
}
`;
